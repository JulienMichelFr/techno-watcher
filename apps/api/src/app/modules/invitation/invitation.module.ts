import { Module } from '@nestjs/common';
import { PrismaModule } from '../prisma/prisma.module';
import { InvitationService } from './services/invitation/invitation.service';

@Module({
  imports: [PrismaModule],
  providers: [InvitationService],
  exports: [InvitationService],
})
export class InvitationModule {}
