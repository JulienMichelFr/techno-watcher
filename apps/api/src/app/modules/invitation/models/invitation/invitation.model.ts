export class InvitationModel {
  public id!: number;
  public code!: string;
  public alreadyUsed!: boolean;
}
