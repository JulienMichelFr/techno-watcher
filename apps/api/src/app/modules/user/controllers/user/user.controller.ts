import { Body, Controller, Post } from '@nestjs/common';
import { Prisma, User } from '@prisma/client';
import { UserService } from '../../services/user/user.service';

@Controller('users')
export class UserController {
  public constructor(private readonly userService: UserService) {}

  @Post()
  public async create(@Body() user: Prisma.UserCreateInput): Promise<User> {
    return this.userService.create(user);
  }
}
