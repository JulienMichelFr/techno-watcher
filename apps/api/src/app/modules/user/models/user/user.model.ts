export class UserModel {
  public id!: number;
  public username!: string;
  public email!: string;
  public password!: string;
  public refreshToken?: string;
}
