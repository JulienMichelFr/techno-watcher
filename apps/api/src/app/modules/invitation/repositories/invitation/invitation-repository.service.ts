import { InvitationModel } from '../../models/invitation/invitation.model';

export abstract class InvitationRepositoryService {
  public abstract findByCode(code: string): Promise<InvitationModel>;
}
