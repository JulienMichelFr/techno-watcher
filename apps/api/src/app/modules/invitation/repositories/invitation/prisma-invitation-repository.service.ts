import { Injectable, NotFoundException } from '@nestjs/common';
import { InvitationRepositoryService } from './invitation-repository.service';
import { PrismaService } from '../../../prisma/prisma.service';
import { Invitation, Prisma, User } from '@prisma/client';
import { InvitationModel } from '../../models/invitation/invitation.model';

export type InvitationAndSelect = Invitation & { user: Pick<User, 'id'> | null };

@Injectable()
export class PrismaInvitationRepositoryService extends InvitationRepositoryService {
  private readonly invitationSelect: Prisma.InvitationSelect = {
    id: true,
    user: {
      select: {
        id: true,
      },
    },
  };

  public constructor(private readonly prisma: PrismaService) {
    super();
  }

  public createInvitationModel(invitation: InvitationAndSelect): InvitationModel {
    const result: InvitationModel = new InvitationModel();
    result.id = invitation.id;
    result.code = invitation.code;
    result.alreadyUsed = !!invitation.user;
    return result;
  }

  public override async findByCode(code: string): Promise<InvitationModel> {
    const result: InvitationAndSelect = (await this.prisma.invitation.findUnique({
      select: this.invitationSelect,
      where: {
        code,
      },
    })) as InvitationAndSelect;

    if (!result) {
      throw new NotFoundException(`Invitation with code ${code} not found`);
    }

    return this.createInvitationModel(result);
  }
}
