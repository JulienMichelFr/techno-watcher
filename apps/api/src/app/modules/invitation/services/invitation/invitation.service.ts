import { Injectable } from '@nestjs/common';
import { InvitationRepositoryService } from '../../repositories/invitation/invitation-repository.service';
import { InvitationModel } from '../../models/invitation/invitation.model';

@Injectable()
export class InvitationService {
  public constructor(private readonly invitationRepository: InvitationRepositoryService) {}

  public async findByCode(code: string): Promise<InvitationModel> {
    return this.invitationRepository.findByCode(code);
  }
}
