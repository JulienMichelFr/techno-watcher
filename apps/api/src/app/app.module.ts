import { Module } from '@nestjs/common';

import { PostModule } from './modules/post/post.module';
import { UserModule } from './modules/user/user.module';
import { AuthModule } from './modules/auth/auth.module';
import { PrismaModule } from './modules/prisma/prisma.module';
import { ConfigModule } from '@nestjs/config';

@Module({
  imports: [ConfigModule.forRoot({ isGlobal: true }), PostModule, UserModule, AuthModule, PrismaModule],
  controllers: [],
  providers: [],
})
export class AppModule {}
