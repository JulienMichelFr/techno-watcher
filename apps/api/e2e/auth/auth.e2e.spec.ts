import { INestApplication, ValidationPipe } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { AppModule } from '../../src/app/app.module';
import { SeederModule } from '../seeders/seeder.module';
import { UserSeeder } from '../seeders/user.seeder';
import { InvitationSeeder } from '../seeders/invitation.seeder';
import * as supertest from 'supertest';
import { AuthResponseModel, RefreshTokenDto, SignInDTO, SignUpDTO } from '@techno-watcher/api-models';
import { signIn } from '../helpers/signin';
import { sign } from 'jsonwebtoken';

describe('Auth', () => {
  let app: INestApplication;
  let userSeeder: UserSeeder;
  let invitationSeeder: InvitationSeeder;

  beforeAll(async () => {
    const moduleRef: TestingModule = await Test.createTestingModule({
      imports: [AppModule, SeederModule],
    }).compile();

    app = moduleRef.createNestApplication();

    app.enableCors();
    app.useGlobalPipes(
      new ValidationPipe({
        whitelist: true,
        transform: true,
      })
    );
    await app.init();
    userSeeder = moduleRef.get<UserSeeder>(UserSeeder);
    invitationSeeder = moduleRef.get<InvitationSeeder>(InvitationSeeder);
    await userSeeder.seed();
  });

  describe('Sign in', () => {
    it('should sign in', async () => {
      const data: SignInDTO = {
        email: userSeeder.getAll()[0].email,
        password: 'Pas$w0rd',
      };
      return supertest(app.getHttpServer())
        .post('/auth/sign-in')
        .send(data)
        .expect(200)
        .expect((response) => {
          expect(response.body as AuthResponseModel).toMatchSnapshot({
            refreshToken: expect.any(String),
            accessToken: expect.any(String),
          });
        });
    });

    it('should reject unknown email', async () => {
      const data: SignInDTO = {
        email: 'invalid@email.com',
        password: 'Pas$w0rd',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-in')
        .send(data)
        .expect(401)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject unknown password', async () => {
      const data: SignInDTO = {
        email: userSeeder.getAll()[0].email,
        password: 'unknown password',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-in')
        .send(data)
        .expect(401)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject invalid email', async () => {
      const data: SignInDTO = {
        email: 'not an email',
        password: 'unknown password',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-in')
        .send(data)
        .expect(400)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });
  });

  describe('Sign up', () => {
    it('should reject invalid email', async () => {
      const data: SignUpDTO = {
        email: 'not an email',
        password: 'Pas$w0rd',
        invitation: invitationSeeder.getAll()[2].code,
        username: 'username',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(400)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject already used email', async () => {
      const data: SignUpDTO = {
        email: userSeeder.getAll()[0].email,
        password: 'Pas$w0rd',
        invitation: invitationSeeder.getAll()[2].code,
        username: 'username',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(400)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject invalid password', async () => {
      const data: SignUpDTO = {
        email: 'user@email.com',
        password: 'invalid password',
        invitation: invitationSeeder.getAll()[2].code,
        username: 'username',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(400)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject invalid invitation code', async () => {
      const data: SignUpDTO = {
        email: 'user@email.com',
        password: 'Pas$w0rd',
        invitation: 'invalid invitation code',
        username: 'username',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(401)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject already used invitation code', async () => {
      const data: SignUpDTO = {
        email: 'user@email.com',
        password: 'Pas$w0rd',
        invitation: invitationSeeder.getAll()[0].code,
        username: 'username',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(401)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject invalid username', async () => {
      const data: SignUpDTO = {
        email: 'user@email.com',
        password: 'Pas$w0rd',
        invitation: invitationSeeder.getAll()[2].code,
        username: 'u',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(400)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should reject already used username', async () => {
      const data: SignUpDTO = {
        email: 'user@email.com',
        password: 'Pas$w0rd',
        invitation: invitationSeeder.getAll()[2].code,
        username: userSeeder.getAll()[0].username,
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(400)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should sign up', () => {
      const data: SignUpDTO = {
        email: 'user@email.com',
        password: 'Pas$w0rd',
        invitation: invitationSeeder.getAll()[2].code,
        username: 'user',
      };

      return supertest(app.getHttpServer())
        .post('/auth/sign-up')
        .send(data)
        .expect(201)
        .expect((response) => {
          expect(response.body as AuthResponseModel).toMatchSnapshot({
            refreshToken: expect.any(String),
            accessToken: expect.any(String),
          });
        });
    });
  });

  describe('Refresh token', () => {
    it('should reject invalid refresh token', async () => {
      const token: string = sign({}, process.env.JWT_SECRET, { expiresIn: '0s' });
      const data: RefreshTokenDto = {
        refreshToken: token,
      };

      return supertest(app.getHttpServer())
        .post('/auth/refresh-token')
        .send(data)
        .expect(403)
        .expect((response) => {
          expect(response.body).toMatchSnapshot();
        });
    });

    it('should refresh token', async () => {
      const { refreshToken } = await signIn(app);
      const data: RefreshTokenDto = {
        refreshToken,
      };

      return supertest(app.getHttpServer())
        .post('/auth/refresh-token')
        .send(data)
        .expect(200)
        .expect((response) => {
          expect(response.body as AuthResponseModel).toMatchSnapshot({
            refreshToken: expect.any(String),
            accessToken: expect.any(String),
          });
        });
    });
  });
});
