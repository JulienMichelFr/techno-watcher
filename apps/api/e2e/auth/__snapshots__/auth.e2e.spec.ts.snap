// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth Refresh token should refresh token 1`] = `
Object {
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
}
`;

exports[`Auth Refresh token should reject invalid refresh token 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Token is expired",
  "statusCode": 401,
}
`;

exports[`Auth Sign in should reject invalid email 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "email must be an email",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Sign in should reject unknown email 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Invalid credentials",
  "statusCode": 401,
}
`;

exports[`Auth Sign in should reject unknown password 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Invalid credentials",
  "statusCode": 401,
}
`;

exports[`Auth Sign in should sign in 1`] = `
Object {
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
}
`;

exports[`Auth Sign up should reject already used email 1`] = `
Object {
  "error": "Bad Request",
  "message": "Could not create resource, email is already used",
  "statusCode": 400,
}
`;

exports[`Auth Sign up should reject already used invitation code 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Invalid invitation",
  "statusCode": 401,
}
`;

exports[`Auth Sign up should reject already used username 1`] = `
Object {
  "error": "Bad Request",
  "message": "Could not create resource, username is already used",
  "statusCode": 400,
}
`;

exports[`Auth Sign up should reject invalid email 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "email must be an email",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Sign up should reject invalid invitation code 1`] = `
Object {
  "error": "Unauthorized",
  "message": "Invalid invitation",
  "statusCode": 401,
}
`;

exports[`Auth Sign up should reject invalid password 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "Password is too weak",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Sign up should reject invalid username 1`] = `
Object {
  "error": "Bad Request",
  "message": Array [
    "username must be longer than or equal to 4 characters",
  ],
  "statusCode": 400,
}
`;

exports[`Auth Sign up should sign up 1`] = `
Object {
  "accessToken": Any<String>,
  "refreshToken": Any<String>,
}
`;
